File: kafpy/fanout.py
"""Fan-out builder for KafPy."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable

from ._kafpy import Consumer

__all__ = ["FanOutBuilder", "FanOutRegistration", "FanInRegistration"]


@dataclass(frozen=True)
class FanInRegistration:
    """Result of a fan-in handler registration.

    Attributes:
        handler_key: The unique identifier for this fan-in handler.
        fan_in_id: Unique numeric ID generated at registration time.
        sources: The list of source topics that feed this handler.
    """

    handler_key: str
    fan_in_id: int
    sources: list[str]


@dataclass(frozen=True)
class FanOutRegistration:
    """Result of a fan-out registration.

    Attributes:
        group_name: The identifier for this fan-out group.
        fan_out_id: Unique ID for this fan-out dispatch (generated at registration time).
        sink_topics: The list of sink topics.
    """

    group_name: str
    fan_out_id: int
    sink_topics: list[str]


@dataclass(frozen=True)
class FanOutBuilder:
    """Builder for configuring fan-out handler registration.

    Use `consumer.register_fanout(...)` to obtain an instance,
    then call `.max_fan_out(n)` to set the max degree, then `.register()`.

    Example::

        builder = consumer.register_fanout(
            group_name="enrichment",
            sink_topics=["topic-a", "topic-b"],
            handler=my_handler,
        )
        registration = builder.max_fan_out(8).register()
    """

    timeout_ms: int | None = None
    consumer: Consumer = field(default=None)
    group_name: str = ""
    sink_topics: list[str] = field(default_factory=list)
    handler: Callable = field(default=None)
    max_fan_out: int | None = None

    def register(self) -> "FanOutRegistration":
        """Register the fan-out group with the consumer.

        Returns:
            FanOutRegistration with group_name, fan_out_id, sink_topics.
        """
        result = self.consumer.register_fanout(
            self.group_name,
            self.sink_topics,
            self.handler,
            self.max_fan_out,
            self.timeout_ms,
        )
        return FanOutRegistration(
            group_name=result.group_name,
            fan_out_id=result.fan_out_id,
            sink_topics=result.sink_topics,
        )

File: src/callbacks/fan_out.rs
//! Fan-out bridge — assembles fan-out config and registers sinks with PyConsumer.
//!
//! Called by `PyConsumer::register_fanout` to attach a fan-out group to one or more
//! sink topics, all sharing the same Python handler callable.

use crate::pyconsumer::PyConsumer;
use crate::python::handler::{HandlerMode, PythonHandler};
use crate::worker_pool::fan_out::FanOutConfig;
use std::sync::Arc;
use std::time::Duration;
use std::sync::atomic::{AtomicU64, Ordering};

/// Global counter for generating unique fan_out_ids.
static FAN_OUT_COUNTER: AtomicU64 = AtomicU64::new(1);

/// Python-facing fan-out builder.
///
/// Constructed from Python via `FanOutBuilderRust::new(...)`, then
/// `.max_fan_out(n)` to configure the degree, then `.register()` to
/// attach all sink handlers to the PyConsumer.
#[derive(Clone)]
pub struct FanOutBuilderRust {
    group_name: String,
    sink_topics: Vec<String>,
    callback: Arc<pyo3::Py<pyo3::PyAny>>,
    mode: HandlerMode,
    timeout_ms: Option<u64>,
    max_fan_out: Option<u8>,
}

impl FanOutBuilderRust {
    /// Create a new fan-out builder.
    ///
    /// - `group_name` — identifier for this fan-out group
    /// - `sink_topics` — list of sink topic names to fan out to
    /// - `callback` — the Python callable invoked for each sink topic
    /// - `mode` — handler execution mode (derived from callback inspection)
    /// - `timeout_ms` — per-branch execution timeout in milliseconds
    pub fn new(
        group_name: String,
        sink_topics: Vec<String>,
        callback: Arc<pyo3::Py<pyo3::PyAny>>,
        mode: HandlerMode,
        timeout_ms: Option<u64>,
    ) -> Self {
        Self {
            group_name,
            sink_topics,
            callback,
            mode,
            timeout_ms,
            max_fan_out: None,
        }
    }

    /// Set the maximum fan-out degree.
    ///
    /// Capped at 64 regardless of input.
    pub fn max_fan_out(mut self, n: u8) -> Self {
        self.max_fan_out = Some(n.min(64));
        self
    }

    /// Register the fan-out group with the PyConsumer.
    ///
    /// Attaches each sink topic as a handler and returns the fan_out_id.
    /// Returns `(group_name, fan_out_id)` for constructing `FanOutRegistration`.
    pub fn register_into_consumer(self, py_consumer: &mut PyConsumer, max: u8) -> (String, u64) {
        let fan_out_id = FAN_OUT_COUNTER.fetch_add(1, Ordering::SeqCst);

        for sink_topic in &self.sink_topics {
            // Build a PythonHandler for this sink topic
            let timeout = self.timeout_ms.map(Duration::from_millis);
            let handler = PythonHandler::new(
                Arc::clone(&self.callback),
                None, // retry_policy — use default from config if needed
                self.mode.clone(),
                None, // batch_policy
                timeout,
                format!("{}[{}]", self.group_name, sink_topic),
                None, // rayon_pool
                None, // middleware
            );

            // Attach fan-out config to the handler so dispatch knows it's a sink
            let fan_out_config = FanOutConfig {
                sinks: Vec::new(), // sinks are registered separately per primary message
                max_fan_out: max,
                slot_manager: None,
            };

            // Register with PyConsumer's handler map directly
            // This bypasses add_handler to attach fan_out config
            py_consumer.add_handler_with_fan_out(
                sink_topic.clone(),
                Arc::new(handler),
                fan_out_config,
            );
        }

        (self.group_name, fan_out_id)
    }
}

/// Extension trait to allow PyConsumer::add_handler to accept fan-out config.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub enum HandlerModeRust {
    #[default]
    SingleSync,
    SingleAsync,
    BatchSync,
    BatchAsync,
    StreamingAsync,
}

impl HandlerModeRust {
    pub fn from_opt_str(s: Option<&str>) -> Self {
        match s {
            Some("async") => HandlerModeRust::SingleAsync,
            Some("batch_sync") => HandlerModeRust::BatchSync,
            Some("batch_async") => HandlerModeRust::BatchAsync,
            Some("streaming_async") => HandlerModeRust::StreamingAsync,
            _ => HandlerModeRust::SingleSync,
        }
    }
}

impl From<HandlerModeRust> for HandlerMode {
    fn from(_: HandlerModeRust) -> Self {
        HandlerMode::SingleSync
    }
}
